xsd/parse.go

```diff
@@ -103,14 +105,48 @@ func Normalize(docs ...[]byte) ([]*xmltree.Element, error) {
 			result = append(result, root.Search(schemaNS, "schema")...)
 		}
 	}
+
+	// Copy element names to anonymous types in order to preserve context within
+	// generated type names. This is done in two passes in order to preserve
+	// as much naming context as possible. Potential collisions are avoided by
+	// appending a suffix to names that would otherwise collide. Suffixes are
+	// based on hashes of the type being renamed in order to decouple generated
+	// type names from the ordering in which the xsd documents were passed as
+	// arguments, providing for more maintainable code based on outputed
+	// GO file
+	namedTypesByNS := map[string]map[xml.Name]string{}
+	namesToBeCopiedByNS := map[string]map[xml.Name][]string{}
+	for _, root := range result {
+		if err := prepCopyEltNamesToAnonTypes(
+			root,
+			namedTypesByNS,
+			namesToBeCopiedByNS,
+		); err != nil {
+			return nil, err
+		}
+	}
 	for _, root := range result {
-		copyEltNamesToAnonTypes(root)
+		copyEltNamesToAnonTypes(
+			root,
+			namedTypesByNS,
+			namesToBeCopiedByNS,
+		)
 	}
+
```

Comment:
As I mentioned in the pull request, using hash based suffixes for anon type names also prevents cascading type renames if an anon type is added early in a document.

Reply:
Thanks!